<a href="https://colab.research.google.com/github/HwangHanJae/ml-definitive-guide-pratice/blob/main/pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

read_csv()의 특징 :
- 필드 구분 문자가 콤마(',')로 구분
- (파일명, sep ='\t')으로 탭으로 구분된 데이터 로드 가능

read_table()의 특징 : 
- 필드 구분 문자가 탭('\t')으로 구분

read_fwf()의 특징 : 
- Fixed Width 고정 길이 기반의 칼럼 포맷

In [ ]:
import pandas as pd

#캐글 - 타이타닉 탑승자 데이터
#read_csv()로 읽어오기, (filepath_of_buffer, sep = ','...)
#filepath_or_buffer 자리에는 파일 경로

file_path = "/content/drive/MyDrive/Colab Notebooks/DataScience/Data/titanic/train.csv"
titanic_df = pd.read_csv(file_path)
print("titalnic_df 변수 type : ", type(titanic_df))

#head()는 defalut는 5, 인자값으로 n을 지정하면 n만큼 n개의 로우를 반환
titanic_df.head()
#index(0,1,3,4...)객체는 sql의 pk와 유사하게 고유의 레코드를 식별하는 역할

titalnic_df 변수 type :  <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [ ]:
#shape : 생성된 DataFrame 객체의 크기를 확인

print("titanic_df 의 크기 : ",titanic_df.shape)
print("생성된 titanic_df는 {}개의 행(로우)과 {}개의 열(칼럼)을 가지고 있습니다.".format(titanic_df.shape[0], titanic_df.shape[1]))

titanic_df 의 크기 :  (891, 12)
생성된 titanic_df는 891개의 행(로우)과 12개의 열(칼럼)을 가지고 있습니다.


In [ ]:
#info() : 총 데이터 건수와 데이터타입, Null 건수를 확인
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


info() 해석
1. RangeIndex는 DataFrame index의 범위를 나타냅니다. 
- 전체 로우의 수를 알 수 있습니다. 데이터는 891개 로우 입니다. 그리고 컬럼수는 12개입니다.

2. Dtype은 칼럼별 타입을 나타냅니다.
- Pclass 칼럼의 데이터 타입은 int64형 입니다.
- Name 칼럼의 데이터 타입은 object입니다. 이는 문자열타입으로 생각해도 무방합니다.

3. Non-Null Count를 보면 몇개의 데이터가 non-null인지 나타냅니다.
- Age 칼럼의 714 non-null의 의미는 Age 칼럼 891개 데이터 중 714개가 Null이 아니며 177개는 Null이라는 의미입니다.

4. dtype : 은 전체 12개의 칼럼들의 타입을 요약한 것 입니다.
- 2개 칼럼이 float64, 5개의 칼럼이 int64, 5개 칼럼이 object 타입입니다.


In [ ]:
#describe() 오직 숫자형(int, float 등)칼럼의 분포도만 조사하며 자동으로 object 타입의 칼럼은 출력에서 제외합니다.
titanic_df.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

In [ ]:

value_counts = titanic_df["Pclass"].value_counts()

print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


In [ ]:
#DataFrame의 []연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 Series 객체를 반환
titanic_pclass = titanic_df["Pclass"]

print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [ ]:
#Series 타입에도 head() 메서드를 사용할 수 있음
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [ ]:
#넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환
import numpy as np

col_name1 =['co1']

#리스트를 이용해 DataFrame 생성
list1 = [1,2,3]
df_list1 = pd.DataFrame(list1, columns=col_name1)
print("1차원 리스트로 만든 DataFrame : \n ", df_list1)

#넘파이 ndarray를 이용해 DataFrame 생성.
array1 = np.array(list1)
df_array1 = pd.DataFrame(array1, columns=col_name1)
print("1차원 ndarray로 만든 DataFrame : \n", df_array1)

1차원 리스트로 만든 DataFrame : 
     co1
0    1
1    2
2    3
1차원 ndarray로 만든 DataFrame : 
    co1
0    1
1    2
2    3


In [ ]:
#2차원 형태의 데이터를 기반으로 DataFrame생성

#3개의 칼럼명이 필요함
col_name2 = ['col1', 'col2','col3']

#2행 x 3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환

list2 = [[1,2,3],
         [11,22,33]]

array2 = np.array(list2)
#array2 의 크기 확인
print("array2 의 크기 : ", array2.shape)
print("array2는 {}행 x {}열의 형태이다. \n".format(array2.shape[0], array2.shape[1]))

#리스트를 이용하여 DataFrame 만들기
df_list2 = pd.DataFrame(data=list2, columns=col_name2)
print("2차원 리스트로 만든 DataFrame : \n", df_list2)

#ndarray를 이용하여 DataFrame 만들기
df_array2 = pd.DataFrame(data=array2, columns=col_name2)
print("2차원 ndarray로 만든 DataFrame : \n", df_array2)

array2 의 크기 :  (2, 3)
array2는 2행 x 3열의 형태이다. 

2차원 리스트로 만든 DataFrame : 
    col1  col2  col3
0     1     2     3
1    11    22    33
2차원 ndarray로 만든 DataFrame : 
    col1  col2  col3
0     1     2     3
1    11    22    33


In [ ]:
#딕셔너리를 DataFrame으로 변환
#딕셔너리의 key는 칼럼, value는 키에 해당하는 칼럼 데이터로 변환

#key는 문자열, value는 리스트형(또는 ndarray)칼럼 데이터로 매핑
dict1 = {'col1' : [1,2,3], "col2" : [11,22,33], "col3" : [111,222,333]}
df_dict = pd.DataFrame(dict1)

print("딕셔너리로 만든 DataFrame : \n", df_dict)

딕셔너리로 만든 DataFrame : 
    col1  col2  col3
0     1    11   111
1     2    22   222
2     3    33   333


In [ ]:
#많은 머신러닝 패키지는 기본 데이터 형으로 ndarray를 사용함
#따라서 DataFrame을 ndarray로 변환할 수 있어야 함
#DataFrame 객체의 values를 이용할 수 있음

#DataFrame을 ndarray로 변환
array3 = df_dict.values
print("df_dict.values 타입 : ", type(array3))
print("df_dict.values의 크기 : ", array3.shape)
#array3을 출력
print(array3)

df_dict.values 타입 :  <class 'numpy.ndarray'>
df_dict.values의 크기 :  (3, 3)
[[  1  11 111]
 [  2  22 222]
 [  3  33 333]]


In [ ]:
df = df_dict.copy()

#DataFrame을 리스트와 딕셔너리로 변환

#DataFrame -> list
#세부 : DataFrame -> ndarray -> list
list3 = df.values.tolist()
print("df.values.tolist()의 타입 : ", type(list3))
print(list3)

#DataFrame -> dict
#to_dict()의 인자값은 'list'로 지정하면 딕셔너리의 value가 list로 변환됨
dict3 = df.to_dict('list')
print("df.values.to_dict()의 타입 : ", type(dict3))
print(dict3)

df.values.tolist()의 타입 :  <class 'list'>
[[1, 11, 111], [2, 22, 222], [3, 33, 333]]
df.values.to_dict()의 타입 :  <class 'dict'>
{'col1': [1, 2, 3], 'col2': [11, 22, 33], 'col3': [111, 222, 333]}


In [ ]:
#DataFrame의 칼럼 데이터 세트 생성과 수정
titanic_df["Age_0"] = 0

titanic_df.head()

PassengerId  Survived  Pclass  ... Cabin Embarked  Age_0
0            1         0       3  ...   NaN        S      0
1            2         1       1  ...   C85        C      0
2            3         1       3  ...   NaN        S      0
3            4         1       1  ...  C123        S      0
4            5         0       3  ...   NaN        S      0

[5 rows x 13 columns]

In [ ]:
#기존 칼럼을 이용하여 새로운 칼럼 생성

#Age를 이용
titanic_df["Age_by_10"] = titanic_df["Age"] * 10

#SibSp와 Parch를 이용
titanic_df["Family_No"] = titanic_df["SibSp"] + titanic_df["Parch"] + 1

#결과 확인
titanic_df.head()

PassengerId  Survived  Pclass  ... Age_0 Age_by_10  Family_No
0            1         0       3  ...     0     220.0          2
1            2         1       1  ...     0     380.0          2
2            3         1       3  ...     0     260.0          1
3            4         1       1  ...     0     350.0          2
4            5         0       3  ...     0     350.0          1

[5 rows x 15 columns]

In [ ]:
#기존 칼럼값 변경
#Age_by_10 변경
titanic_df["Age_by_10"] = titanic_df["Age_by_10"] + 100
titanic_df.head()

PassengerId  Survived  Pclass  ... Age_0 Age_by_10  Family_No
0            1         0       3  ...     0     320.0          2
1            2         1       1  ...     0     480.0          2
2            3         1       3  ...     0     360.0          1
3            4         1       1  ...     0     450.0          2
4            5         0       3  ...     0     450.0          1

[5 rows x 15 columns]

drop() 메서드의 원형 :
- DataFrame.drop(**labels**=None, **axis=0**, index=None, columns=None, level=None, **inplace=False**, errors ='raise')
  - labels : 원하는 칼럼명
  - axis : 0은 로우 방향축, 1은 칼럼 방향 축
  - inplace : True면 원본 DataFrame 삭제

In [ ]:
#DataFrame 데이터 삭제

#drop() 메서드를 이용합니다., inplace = False

titanic_drop_df = titanic_df.drop("Age_0", axis=1)
titanic_drop_df.head()

PassengerId  Survived  Pclass  ... Embarked Age_by_10  Family_No
0            1         0       3  ...        S     320.0          2
1            2         1       1  ...        C     480.0          2
2            3         1       3  ...        S     360.0          1
3            4         1       1  ...        S     450.0          2
4            5         0       3  ...        S     450.0          1

[5 rows x 14 columns]

In [ ]:
drop_result = titanic_df.drop(['Age_0', "Age_by_10", "Family_No"], axis=1, inplace=True)
print("inplce = True 로 drop 후 반환된 값 : ", drop_result)
titanic_df.head()

inplce = True 로 drop 후 반환된 값 :  None


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [ ]:
#axis=0 을 사용하여 index 0,1,2(맨 앞 3개 데이터) 로우를 삭제
print("####before drop axis=0####")
print(titanic_df.head())
print("####after drop axis=0####")
print(titanic_df.drop([0,1,2], axis=0).head())

####before drop axis=0####
   PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]
####after drop axis=0####
   PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S
5            6         0       3  ...   8.4583   NaN         Q
6            7         0       1  ...  51.8625   E46         S
7            8         0       3  ...  21.0750   NaN         S

[5 rows x 12 columns]


In [ ]:
#인덱스 객체
#DataFrame, Seires 데이터에서 Index 객체만 추출
indexes = titanic_df.index
print("Index 객체 : ",indexes)
#index 객체를 실제 값 array로 변환
print("indexes.values 타입  : ", type(indexes.values))
print("Index 객체 array값 : \n", indexes.values)

Index 객체 :  RangeIndex(start=0, stop=891, step=1)
indexes.values 타입  :  <class 'numpy.ndarray'>
Index 객체 array값 : 
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217

In [ ]:
#Index객체는 단일 값 반환 및 슬라이싱 가능
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


In [ ]:
#Index 객체는 함부로 변경할 수 없음
indexes[0] = 5
#Index does not support mutable operations

TypeError: ignored

In [ ]:
#Series 객체는 Index 객체를 포함하지만 Series 객체에 연산 함수를 적용할 때
#Index는 연산에서 제외됩니다.

series_fair = titanic_df["Fare"]
print("Fair Series max 값 : ", series_fair.max())
print("Fair Series min 값 : ,", series_fair.min())
print("sum() Fair Series : ", series_fair.sum())
print("Fair Series + 3 :\n", (series_fair + 3).head(3))

Fair Series max 값 :  512.3292
Fair Series min 값 : , 0.0
sum() Fair Series :  28693.9493
Fair Series + 3 :
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


In [ ]:
#reset_index() : 새롭게 인덱스를 연속 숫자 형으로 할당
#기존 인덱스는 'index'라는 새로운 칼럼 명으로 추가
titanic_reset_df = titanic_df.reset_index()
titanic_reset_df.head()

index  PassengerId  Survived  ...     Fare Cabin Embarked
0      0            1         0  ...   7.2500   NaN        S
1      1            2         1  ...  71.2833   C85        C
2      2            3         1  ...   7.9250   NaN        S
3      3            4         1  ...  53.1000  C123        S
4      4            5         0  ...   8.0500   NaN        S

[5 rows x 13 columns]

In [ ]:
#Series.reset_index()를 하게 되면 DataFrame으로 반환
print("#### before reset_index ####")
value_counts = titanic_df["Pclass"].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입 : ', type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print("#### after reset_index ####")
print(new_value_counts)
print('new_value_counts 객체 변수 타입 : ', type(new_value_counts))

#### before reset_index ####
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입 :  <class 'pandas.core.series.Series'>
#### after reset_index ####
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입 :  <class 'pandas.core.frame.DataFrame'>


In [ ]:
print("#### before new_reset_index drop = False ####")
new_reset_counts = value_counts.reset_index(inplace=False,drop=False)
print("new_value_counts 객체 변수 타입 : ", type(new_reset_counts))
print(new_reset_counts)
print("#### after new_reset_index drop = True ####")
drop_reset_counts = value_counts.reset_index(drop=True)
print('drop_reset_counts 객체 변수 타입 : ', type(drop_reset_counts))
print(drop_reset_counts)

#### before new_reset_index drop = False ####
new_value_counts 객체 변수 타입 :  <class 'pandas.core.frame.DataFrame'>
   index  Pclass
0      3     491
1      1     216
2      2     184
#### after new_reset_index drop = True ####
drop_reset_counts 객체 변수 타입 :  <class 'pandas.core.series.Series'>
0    491
1    216
2    184
Name: Pclass, dtype: int64
